hoomd/pytest/test_gsd.py

```diff
@@ -0,0 +1,378 @@
+                                 mode='wb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    sim.run(3)
+
+    with gsd.hoomd.open(name=filename, mode='rb') as traj:
+        for step in range(3):
+            assert traj[step].particles.N == 0
+
+
+def test_write_gsd_type_filter(simulation_factory, device,
+                               hoomd_snapshot, tmp_path):
+
+    # test using Type filter
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 filter=hoomd.filter.Type(['p2']),
+                                 mode='wb')
+    sim.operations.writers.append(gsd_writer)
+
+    sim.run(5)
+    snapshot = sim.state.snapshot
+
+    with gsd.hoomd.open(name=filename, mode='rb') as traj:
+        for step in range(5):
+            np.testing.assert_array_equal(traj[step].particles.typeid,
+                                          [1] * int(snapshot.particles.N / 2))
+
+
+def test_write_gsd_truncate(simulation_factory, device,
+                            hoomd_snapshot, tmp_path):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 truncate=True,
+                                 mode='wb')
+    sim.operations.writers.append(gsd_writer)
+
+    snapshot_list = []
+    for _ in range(2):
+        sim.run(1)
+        snapshot_list.append(sim.state.snapshot)
+
+    with gsd.hoomd.open(name=filename, mode='rb') as traj:
+        assert_equivalent_snapshots(traj[0], snapshot_list[-1])
+        assert_equivalent_snapshots(traj[-1], snapshot_list[-1])
+
+
+def test_write_gsd_dynamic(simulation_factory, device,
```

Comment:
We should test that all the dynamic options work, and that when they are not chosen that the corresponding values don't get written. Also, this would likely be cleaner using a parametrized fixture or `pytest.mark.parametrize`.

Reply:
Or more specifically, check that when the dynamic option is not enabled, only dynamic quantities are written, static ones are only written in the first frame? In the test I wrote, would velocity still be stored in every frame even if the ```dynamic =None```?

Reply:
I think testing that only `property` attributes get written by default, and when another `dynamic` option is chosen that it adds the correct fields, and when it is not chosen that those fields are not written.

Reply:
I'm a little confused about what the doc says, so if ```dynamic``` is set to default, only ```property``` attributes are written and not other attributes that are also changing in each frame? If another```dynamic``` option is chosen, the corresponding fields will be added, but still if one of the newly added attributes is static, it's written only in the first frame?

Reply:
For example, say you have a simulation where positions and velocities change every frame, but particle orientations stay [1, 0, 0, 0].
With `dynamic=['property']`: `position` and `velocity` will be written on the first frame. When you read the file, `position` will be read for each frame. When you read the file and access `velocity` at frame i (i != 0), GSD will provide the velocity from the first frame. This option exists to save disk space. Users need to opt-in to writing out additional quantities every frame.

With `dynamic=['property', 'momentum']`: `position` and `velocity` will be written on every frame. When you read the file, `position` and `velocity` will both be read for each frame.

In both cases, `orientation` will not be written at all. GSD understands that the default orientation is [1, 0, 0, 0] and will provide that per-particle value on read even though the data is not in the file. This behavior is to save disk space.

Reply:
I still have some questions about testing ```dynamic```, especially about ```dynamic=['attribute']``` and ```dynamic=['topology']```. For these, do we need to make simulations where the corresponding quantities change to test them? And for the cases where the quantities are only stored in the first frame, do we need to test that they're not written in every frame, but rather read from the first frame?

Reply:
Ideally, yes. You would need to change e.g. the bonds in the system or particle types from one frame to the next in order to test that these dynamic flags function. The GSD python library does not indicate whether a particular quantity is given in every frame or read from the fist frame. Instead, you have to inspect the frame and see if it hast the expected contents (e.g. that the typeid still match those set in frame 0 even though you are looking at frame 1).

Instead of using `assert_equivalent_snapshots` for this, you can use targeted assert checks to verify the expected data. e.g.
```
# attributes static
assert traj[0].typeid == [0, 0]
assert traj[1].typeid == [0, 0]

# attributes dynamic
assert traj[0].typeid == [0, 0]
assert traj[1].typeid == [1, 0]
```